pkg/controllers/rollout/controller_integration_test.go

```diff
@@ -0,0 +1,377 @@
+	"fmt"
+	"strconv"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+)
+
+const (
+	timeout  = time.Second * 10
+	duration = time.Second * 20
+	interval = time.Millisecond * 250
+)
+
+var testCRPName string
+
+var _ = Describe("Test the rollout Controller", func() {
+
+	var bindings []*fleetv1beta1.ClusterResourceBinding
+	var resourceSnapshots []*fleetv1beta1.ClusterResourceSnapshot
+	var rolloutCRP *fleetv1beta1.ClusterResourcePlacement
+
+	BeforeEach(func() {
+		testCRPName = "crp" + utils.RandStr()
+		bindings = make([]*fleetv1beta1.ClusterResourceBinding, 0)
+		resourceSnapshots = make([]*fleetv1beta1.ClusterResourceSnapshot, 0)
+	})
+
+	AfterEach(func() {
+		By("Deleting ClusterResourceBindings")
+		for _, binding := range bindings {
+			Expect(k8sClient.Delete(ctx, binding)).Should(SatisfyAny(Succeed(), utils.NotFoundMatcher{}))
+		}
+		By("Deleting ClusterResourceSnapshots")
+		for _, resourceSnapshot := range resourceSnapshots {
+			Expect(k8sClient.Delete(ctx, resourceSnapshot)).Should(SatisfyAny(Succeed(), utils.NotFoundMatcher{}))
+		}
+		By("Deleting ClusterResourcePlacement")
+		Expect(k8sClient.Delete(ctx, rolloutCRP)).Should(SatisfyAny(Succeed(), utils.NotFoundMatcher{}))
+	})
+
+	It("Should rollout all the selected bindings as soon as they are created", func() {
+		// create CRP
+		var targetCluster int32 = 10
+		rolloutCRP = clusterResourcePlacementForTest(testCRPName, createPlacementPolicyForTest(fleetv1beta1.PickNPlacementType, targetCluster))
+		Expect(k8sClient.Create(ctx, rolloutCRP)).Should(Succeed())
+		// create master resource snapshot that is latest
+		masterSnapshot := generateResourceSnapshot(rolloutCRP.Name, 0, true)
+		Expect(k8sClient.Create(ctx, masterSnapshot)).Should(Succeed())
+		By(fmt.Sprintf("master resource snapshot  %s created", masterSnapshot.Name))
+		// create scheduled bindings for master snapshot on target clusters
+		clusters := make([]string, targetCluster)
+		for i := 0; i < int(targetCluster); i++ {
+			clusters[i] = "cluster-" + utils.RandStr()
+			binding := generateClusterResourceBinding(fleetv1beta1.BindingStateScheduled, masterSnapshot.Name, clusters[i])
+			Expect(k8sClient.Create(ctx, binding)).Should(Succeed())
+			By(fmt.Sprintf("resource binding  %s created", binding.Name))
+			bindings = append(bindings, binding)
+		}
+		// check that all bindings are scheduled
+		Eventually(func() bool {
+			for _, binding := range bindings {
+				err := k8sClient.Get(ctx, types.NamespacedName{Name: binding.GetName()}, binding)
+				if err != nil {
+					return false
+				}
+				if binding.Spec.State != fleetv1beta1.BindingStateBound || binding.Spec.ResourceSnapshotName != masterSnapshot.Name {
+					return false
```

Comment:
compare the whole binding spec?

Reply:
address this one?

Reply:
the rest of the spec are dummy. I am sure there is value to compare the rest